File: backend/services/otx_service.py
import requests
import os
from datetime import datetime, timedelta
from dotenv import load_dotenv

load_dotenv()

class OTXService:
    BASE_URL = "https://otx.alienvault.com/api/v1"
    
    def __init__(self):
        self.api_key = os.getenv("OTX_API_KEY")
        self.headers = {
            "X-OTX-API-KEY": self.api_key,
            "Content-Type": "application/json"
        }

    def get_pulses(self, limit=10):
        """Fetch latest threat pulses from OTX."""
        try:
            # Use search/pulses with a generic query to ensure results
            # optimized for freshness
            seven_days_ago = (datetime.now() - datetime.timedelta(days=7)).isoformat()
            url = f"{self.BASE_URL}/search/pulses"
            params = {
                "q": "malware", 
                "limit": limit, 
                "sort": "-modified",
                "modified_since": seven_days_ago
            }
            response = requests.get(url, headers=self.headers, params=params)
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"Error fetching pulses: {e}")
            return None

    def get_pulse_indicators(self, pulse_id):
        """Fetch indicators for a specific pulse."""
        try:
            url = f"{self.BASE_URL}/pulses/{pulse_id}/indicators"
            response = requests.get(url, headers=self.headers)
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"Error fetching indicators for pulse {pulse_id}: {e}")
            return None

File: backend/services/otxService.js
const axios = require('axios');
const path = require('path');
require('dotenv').config({ path: path.resolve(__dirname, '../.env') });

const API_KEY = process.env.OTX_API_KEY;
const BASE_URL = 'https://otx.alienvault.com/api/v1';

const getIpPulses = async (ip) => {
    try {
        const response = await axios.get(`${BASE_URL}/indicators/IPv4/${ip}/general`, {
            headers: {
                'X-OTX-API-KEY': API_KEY
            }
        });
        return response.data;
    } catch (error) {
        console.error('Error fetching OTX pulses for IP:', error.message);
        return null; // Return null on error so the caller knows it failed
    }
};

// Function to search for pulses - useful for gathering general threat data
const searchPulses = async (query, limit = 10) => {
    try {
        const response = await axios.get(`${BASE_URL}/search/pulses`, {
            params: {
                q: query,
                limit: limit
            },
            headers: {
                'X-OTX-API-KEY': API_KEY
            }
        });
        return response.data.results;
    } catch (error) {
        console.error('Error searching OTX pulses:', error.message);
        return [];
    }
}

module.exports = { getIpPulses, searchPulses };
